### PR TITLE
pd: cancel unused region heartbeat rpc

### PR DESCRIPTION
**Pull request by overvenus**
PdClient may leak gRPC calls if PD leader changes rapidly.

Cc #2669

### CODE REVIEW

**Comment by overvenus** [comment]
A friendly ping.

**Comment by BusyJay** [comment]
Can this be solved by just adding keep alive ping?

**Comment by overvenus** [comment]
No, it can't.

File: tests/pd/test_rpc_client.rs [hunk 1]
```diff
@@ -372,3 +372,63 @@ fn test_change_leader_async() {
...
+        .spawn(client.region_heartbeat(region.clone(), peer.clone(), stat.clone()))
+        .forget();
+    rx.recv_timeout(LeaderChange::get_leader_interval())
+        .unwrap();
+
+    let change_leader = |count| {
+        let mut leader = client.get_leader();
+        for _ in 0..count {
+            loop {
+                client.get_region_by_id(1).wait().ok();
```
Comment (BusyJay) [inline tests/pd/test_rpc_client.rs]: This is confusing, use `let _ = xxx` instead.

File: tests/pd/test_rpc_client.rs [hunk 2]
```diff
@@ -372,3 +372,63 @@ fn test_change_leader_async() {
...
+    let region = metapb::Region::new();
+    let peer = metapb::Peer::new();
+    let stat = RegionStat::default();
+    poller
+        .spawn(client.region_heartbeat(region.clone(), peer.clone(), stat.clone()))
+        .forget();
+    rx.recv_timeout(LeaderChange::get_leader_interval())
+        .unwrap();
+
+    let change_leader = |count| {
```
Comment (BusyJay) [inline tests/pd/test_rpc_client.rs]: Ambiguous name. The closure doesn't change leader by itself.

**Comment by overvenus** [comment]
PTAL

File: tests/pd/test_rpc_client.rs [hunk 3]
```diff
@@ -372,3 +372,63 @@ fn test_change_leader_async() {
...
+            }
+        }
+        poller
+            .spawn(client.region_heartbeat(region.clone(), peer.clone(), stat.clone()))
+            .forget();
+        rx.recv_timeout(LeaderChange::get_leader_interval())
+            .unwrap();
+    };
+
+    // Change PD leader once.
```
Comment (BusyJay) [inline tests/pd/test_rpc_client.rs]: Stale comments.

**Comment by overvenus** [comment]
PTAL

**Comment by BusyJay** [comment]
LGTM

File: src/pd/client.rs [hunk 4]
```diff
@@ -307,33 +307,38 @@ impl PdClient for RpcClient {
...
-                        sender
-                            .sink_map_err(Error::Grpc)
-                            .send_all(rx.map_err(|e| {
-                                Error::Other(box_err!("failed to recv heartbeat: {:?}", e))
-                            }).map(|r| (r, WriteFlags::default())))
-                            .map(|(mut sender, _)| sender.get_mut().cancel()),
-                    ) as PdFuture<_>
-                }
-                None => unreachable!(),
+            if let Either::Right(ref sender) = inner.hb_sender {
```
Comment (huachaohuang) [inline src/pd/client.rs]: This change is just a refactor, right?

File: src/pd/util.rs [hunk 5]
```diff
@@ -177,6 +176,13 @@ impl LeaderClient {
         {
             let mut inner = self.inner.wl();
             let (tx, rx) = client.region_heartbeat().unwrap();
+            warn!("heartbeat sender and receiver are stale, refreshing..");
+
+            // Try to cancel an unused heartbeat sender.
+            if let Either::Left(Some(ref mut r)) = inner.hb_sender {
```
Comment (huachaohuang) [inline src/pd/util.rs]: So, the bug is that if we reconnect twice before the previous sender is used, then the previous sender will not be canceled, right?

Comment (overvenus) [reply src/pd/util.rs]: Yes, you are right.

File: src/pd/client.rs (hunk 4 above)
Comment (overvenus) [reply src/pd/client.rs]: Yes, it is.

**Pull request merged**